support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -54,6 +55,53 @@ object GiftRecipient {
     ))
 }
 
+case class GiftCode private(value: String) extends AnyVal
+
+object GiftCode {
+
+  def apply(value: String): Option[GiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GiftCode(_).toRight("invalid gift code"))
+
+}
+
+sealed trait GiftPurchase {
+  def asDigiSub: Option[DigitalSubGiftPurchase] = this match { case a: DigitalSubGiftPurchase => Some(a); case _ => None }
+  def giftRecipient: GiftRecipient
+}
+object GiftPurchase {
+
+  case class WeeklyGiftPurchase(giftRecipient: WeeklyGiftRecipient) extends GiftPurchase
+  case class DigitalSubGiftPurchase(
+    giftRecipient: DigitalSubGiftRecipient,
+    giftCode: GiftCode,
+  ) extends GiftPurchase
+
+  val circeDiscriminator = new CirceDiscriminator("giftPurchaseType")
+  private val discriminatorNonDigiSub = "NonDigiSub"
```

Comment:
Can we make this `Weekly` and the other one `DigitalSub` or `DigitalSubscription`? 
I'd prefer to avoid abbreviations where possible or at least restrict them to one standard one abbreviation rather than mixing between DigitalSubscription/DigitalSub/DigiSub

Reply:
good idea, it's good to be consistent, or drop words completely, rather than random contractions

Reply:
I should also change the existing one too on line 39 of this same file
https://github.com/guardian/support-frontend/blob/d150942050cec1f665e022eb3dae994e337ac7fe/support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala#L38